pandas/tests/frame/test_asof.py

```diff
@@ -75,9 +74,34 @@ def test_missing(self):
         result = df.asof('1989-12-31')
 
         expected = Series(index=['A', 'B'], name=Timestamp('1989-12-31'))
-        assert_series_equal(result, expected)
+        tm.assert_series_equal(result, expected)
 
         result = df.asof(to_datetime(['1989-12-31']))
         expected = DataFrame(index=to_datetime(['1989-12-31']),
                              columns=['A', 'B'], dtype='float64')
-        assert_frame_equal(result, expected)
+        tm.assert_frame_equal(result, expected)
+
+    def test_all_nans(self):
+        # GH 15713
+        # DataFrame is all nans
+        result = DataFrame([np.nan]).asof([0])
+        expected = DataFrame([np.nan])
+        tm.assert_frame_equal(result, expected)
+
+        # testing non-default indexes, multiple inputs
+        dates = date_range('1/1/1990', periods=self.N * 3, freq='25s')
+        result = DataFrame(np.nan, index=self.rng, columns=['A']).asof(dates)
+        expected = DataFrame(np.nan, index=dates, columns=['A'])
+        tm.assert_frame_equal(result, expected)
+
+        # testing multiple columns
+        dates = date_range('1/1/1990', periods=self.N * 3, freq='25s')
+        result = DataFrame(np.nan, index=self.rng, columns=['A', 'B', 'C']).asof(dates)
+        expected = DataFrame(np.nan, index=dates, columns=['A', 'B', 'C'])
+        tm.assert_frame_equal(result, expected)
+
+        # testing scalar input
+        date = date_range('1/1/1990', periods=self.N * 3, freq='25s')[0]
+        result = DataFrame(np.nan, index=self.rng, columns=['A']).asof(date)
+        expected = DataFrame(np.nan, index=[date], columns=['A'])
+        tm.assert_frame_equal(result, expected)
```

Comment:
What you added is not the correct result I think. It are the original columns that are the index of the resulting series, not ``[where]``. The ``where`` does become the name of the Series (i.e. as if you access a row from the dataframe)

Can you add the example above (but then with NaNs instead of the random data) as a test case? The it is really clear what the expected behaviour is.